### PR TITLE
Improve Chat Switch timing in Grafana so it waits for all messages to layout

### LINKED ISSUE #2683: Chat Switch timing does not accurately reflect the test we want to perform
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Grafana data should reflect the total time it takes to tap on a chat and view the messages in that chat

## Actual Result:
Grafana data only reflects the time it takes for a report to mount and the contents still takes some time to render

## Action Performed:
1. Switch to a chat
2. See that the timing log fires well before the chat messages are loaded

## Workaround:
None

## Platform:
 Web
 iOS
 Android
 Desktop App
 Mobile Web

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

### PR DESCRIPTION
**Pull request by marcaaron**
### Details
While A/B testing changes to Onyx cache stuff we discovered that this test does not actually measure the time it takes to tap on a chat and see the entire contents of the chat rendered. What we were testing is the time it takes the `ReportActionsView` to mount, but it might have no chat history available to it yet.

This new test will time how long it takes for all the items to render by debouncing a call passed to the `onLayout` of the view enclosing the chat content so it will record the "end" event once no one has called it after 1 second. We'll also then need to offset the timing by 1 second so that we get the true result.

This will give us a better sense of when the report is "ready to view".

**Note:** This test has one downside at the moment which is that @kidroca has pointed out that `VirtualizedList` may be configured incorrectly and rendering many more items than necessary see issue [here](https://github.com/Expensify/Expensify.cash/issues/2629). Still I think it's valuable to do this test now so we can understand when a report is actually done doing it's "work" to render the UI.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify.cash/issues/2683

### Tests
1. Switch to a chat
2. Wait for the chat messages to load
3. Verify the timing log eventually shows with the time it took for all chat messages to render

### QA Steps
No QA

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### CODE REVIEW

File: src/pages/home/report/ReportActionItem.js
Comment (Julesssss) [inline src/pages/home/report/ReportActionItem.js:128]: Oh nice, didn't know that View has an `onLayout` prop.

**Review by Julesssss — DISMISSED** [review]
LGTM. Will keep in mind for future render timing.

**Comment by Julesssss** [comment]
All yours @MonilBhavsar

File: src/CONST.js
Comment (Julesssss) [inline src/CONST.js:67]: NAB, but is it true that a slow child component taking longer than the denounce time to render would trigger a false measurement?

**Comment by Julesssss** [comment]
> @Julesssss, the code looks good to me.
> You can merge after your conversation

Thanks @MonilBhavsar. Don't forget to approve the PR if you are happy with it, else the review won't count towards your total and you'll be auto-assigned more PRs for review :)

Also just to let you know, NAB == 'Not A Blocker'. You might see people adding `NAB` if a particular comment or change request is not essential and should not block the merge.

**Comment by MonilBhavsar** [comment]
Yes sure, Approving it, From today only I got PR Reviews and I got a **lot**
and NAB makes sense now

**Review by MonilBhavsar — DISMISSED** [review]

**Comment by Julesssss** [comment]
I'd like to run the test once more, but unfortunately ran out of time this afternoon. Will make this my first task tomorrow.

File: src/CONST.js
Comment (marcaaron) [reply src/CONST.js:67]: Yes, I think if we had a child that takes more than a second to render we'd definitely get a funky measurement.

I think it's pretty unlikely for that to happen.

I came across another issue here though that I'm going to look into addressing now.

If we switch chats before thing are done rendering it's possible to get a negative value recorded so I'll look into how to prevent this before proceeding.

File: src/pages/home/report/ReportActionItem.js
Comment (marcaaron) [reply src/pages/home/report/ReportActionItem.js:128]: https://reactnative.dev/docs/view#onlayout

The docs do say

>This event is fired immediately once the layout has been calculated, but the new layout may not yet be reflected on the screen at the time the event is received, especially if a layout animation is in progress.

I think this is maybe as close as we can get to treating an individual chat message as "rendered". There might be a better indicator to use.

Comment (kidroca) [reply src/pages/home/report/ReportActionItem.js:128]: Hey I thought that as well, I don't mean to get you sidetracked but here's what I've found looking for an alternative:

[viewabilityConfigCallbackPairs](https://reactnative.dev/docs/virtualizedlist#viewabilityconfigcallbackpairs)
> List of ViewabilityConfig/onViewableItemsChanged pairs. A specific onViewableItemsChanged will be called when its corresponding ViewabilityConfig's conditions are met. See ViewabilityHelper.js for flow type and further documentation.

And in [ViewabilityHelper.js](https://github.com/facebook/react-native/blob/dc80b2dcb52fadec6a573a9dd1824393f8c29fdc/Libraries/Lists/ViewabilityHelper.js#L34-L40)
> Minimum amount of time (in milliseconds) that an item must be physically viewable before the
viewability callback will be fired.

I've never used this but it seems you can configure this and wait for the callback to get called for all the items or just the first 5-10 (Thought what about when there are less than 10...)

Comment (kidroca) [reply src/pages/home/report/ReportActionItem.js:128]: Anyway I think `onLayout` is close enough maybe 1-2 frames before the item is rendered so that like 16-32ms which is probably too little to care about

File: src/CONST.js
Comment (marcaaron) [reply src/CONST.js:67]: Ok figured that one out and we just need to cancel the debounced method in `componentWillUnmount()` so it does not incorrectly set the timer end when switching to the next chat.

So, do you think it's OK to have this... or should we perhaps increase the time?

>a slow child component taking longer than the denounce time to render would trigger a false measurement

File: src/pages/home/report/ReportActionItem.js
Comment (marcaaron) [reply src/pages/home/report/ReportActionItem.js:128]: Thanks, that's interesting! The docs are kind of sparse on that stuff so I'm not too sure if I'll go down this road right now. I think it will maybe be slightly more complicated to use this then to do the naive idea I had.

Comment (marcaaron) [reply src/pages/home/report/ReportActionItem.js:128]: Mostly because `onViewableItemsChanged` is called not just when the items first appear, but when things scroll etc. And I don't want to think about it 😄

**Review by Julesssss — DISMISSED** [review]

File: src/CONST.js
Comment (Julesssss) [reply src/CONST.js:67]: As unlikely as this is, I think it would be better to set a slightly higher debounce time -- _just in case_. But it's NAB. I'll leave it up to you 👍

**Comment by Julesssss** [comment]
I'm leaving the denounce time increase up to @marcaaron, feel free to merge.

**Review by MonilBhavsar — APPROVED** [review]

**Review by Julesssss — APPROVED** [review]

**Pull request merged**
